code/controllers/subsystem/pathfinding.dm

```diff
@@ -0,0 +1,379 @@
+			new_cost = current[NODE_COST] + current_distance; \
+			reverse_dir_of_expand = REVERSE_DIR(dir); \
+			if(expand) { \
+				expand[NODE_DIR] = expand[NODE_DIR] & ((NORTH|SOUTH|EAST|WEST) ^ reverse_dir_of_expand); \
+				if(new_cost < expand[NODE_COST]) { \
+					if(call(current_turf, can_cross_proc)(caller, expand_turf, ID, dir, reverse_dir_of_expand)) { \
+						expand[NODE_PREVIOUS] = current; \
+						expand[NODE_COST] = new_cost; \
+						expand[NODE_WEIGHT] = new_cost + expand[NODE_HEURISTIC] * PATHFINDING_HEURISTIC_TIEBREAKER_WEIGHT; \
+						expand[NODE_DEPTH] = current[NODE_DEPTH] + 1; \
+						open -= expand; \
+						INJECT_NODE(open, expand); \
+					}; \
+				}; \
+			}; \
+			else { \
+				if(call(current_turf, can_cross_proc)(caller, expand_turf, ID, dir, reverse_dir_of_expand)) { \
+					CALCULATE_DISTANCE(expand_turf, end); \
+					SETUP_NODE(open, T, new_cost, current_distance, current[NODE_DEPTH] + 1, (NORTH|SOUTH|EAST|WEST)^reverse_dir_of_expand, expand_turf); \
+				}; \
+			}; \
+		}; \
+	};
+
+/**
+  * Runs a pathfind with normal A*
+  * For 99.99% of applications, you probably want JPS, seen above.
+  * The reason A* is kept is because it's more likely to return a more optimized path.
+  * Plus, in the far unknown future because I am so, so sure someone will give a care about this, you can implement turf movement costs.
+  * However, for the most part, yeah haha nah, use JPS, not worth the CPU cost.
+  *
+  * @params
+  * * caller - What called this. Can be null, datum, atom, whatever.
+  * * start - turf to start on. If not set, defaults to caller's turf if it's an atom. If neither are set, the proc crashes.
+  * * end - turf to path to.
+  * * can_cross_proc - proc to call on turfs to find if we can pass from it to another turf. /turf/proc/procname, called with arguments(caller, turf/trying_to_reach, obj/item/card/id/ID, dir_to_them, dir_from_them)
+  * * heuristic_type - heuristic type of distance/cost calculations see [code/__DEFINES/pathfinding.dm]
+  * * max_node_depth - maximum depth of nodes to search. INFINITY for infinite.
+  * * max_path_distance - maximum length of returned path using given heuristic can be. 0 for infinite.
+  * * min_target_distance - minimum distance to target to terminate pathfinding. Used to get close to a target rather than to it.
+  * * turf_blacklist_typecache - blacklist typecache of turfs we can't cross no matter what. defaults to space tiles.
+  * * queue - queue to put this in/use with this.
+  * * ID - obj/item/card/id to provide access. why this uses an id card and not an access list, .. don't ask.
+  */
+/datum/controller/subsystem/pathfinding/proc/AStar_pathfind(caller, turf/start, turf/end, can_cross_proc = /turf/proc/pathfinding_can_cross, heuristic_type = PATHFINDING_HEURISTIC_BYOND, max_node_depth = 30, max_path_distance = 0, min_target_distance = 0, turf_blacklist_typecache = SSpathfinding.space_type_cache, queue = PATHFINDING_QUEUE_DEFAULT, obj/item/card/id/ID)
```

Comment:
the two x_pathfind procs are very obviously copypaste, wondering if I should make it one proc with an argument for the kind of pathfinding to use tbh
same arguments too sooo that'd be pretty easy to do.

Reply:
i don't personally think it's that bad right now but if we ever get another algorithm or whatever the copypaste can get silly soo yeah should i do that y/n